### PR TITLE
[Misc] Refactor linear layer weight loading; introduce `BasevLLMParameter` and `weight_loader_v2`

### PR DESCRIPTION
**Pull request by dsikka**
# Summary
- Introduce a series of new parameters to handle the different weight loading cases for parameters loaded to linear layers. 
- Introduces: 

1. `BasevLLMParameter`
2. `ModelWeightParameter`
3. `GroupQuantScaleParameter`
4. `ChannelQuantScaleParameter`
5. `PerTensorScaleParameter`
6. `PackedvLLMParameter`

- Each of these parameters handle the weight-loading logic specific to the different `LinearBase` classes
- Significantly cleans up the `weight_loader` method in each of the `LinearBase` classes
- For now, changes are only made to `compressed-tensors` quantization configs by adding a `weight_loader_v2` method to each of the `LinearBase` classes.  All other quantization configurations are still using the original weight loader, as part of the scope of this PR

## FOLLOW UP:
* convert other integrations to use this framework once design is approved

### CODE REVIEW

File: vllm/model_executor/parameter.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+        offset = sum(self.logical_widths[:shard_id])
+        size = self.logical_widths[shard_id]
+        # update loaded weight with copies for broadcast.
+        loaded_weight = loaded_weight.repeat(size)
+        return param_data[offset:offset + size], loaded_weight
+
+
+class PackedvLLMParameter(vLLMParameter):
+
+    def __init__(self,
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/parameter.py:130]: Can you add a docstring explaining what a `PackedParameter` is and perhaps give an example?

- E.g. explain `AutoGPTQ` puts 4 bit weights packed in int32 datatype.
- Adds complexity because we have to handle indexing for fused loading

File: vllm/model_executor/__init__.py [hunk 2]
```diff
@@ -1,3 +1,4 @@
+from vllm.model_executor.parameter import *
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/__init__.py:1]: please avoid * imports

File: vllm/model_executor/layers/linear.py [hunk 3]
```diff
@@ -471,6 +493,81 @@ def weight_loader(self,
...
+                shard_offset=shard_offset, shard_size=shard_size)
+
+        param_data = param_data.narrow(param.output_dim, shard_offset,
+                                       shard_size)
+        loaded_weight.narrow(param.output_dim, tp_rank * shard_size,
+                             shard_size)
+
+        return param_data, loaded_weight
+
+    def _load_no_shard_id(self, param: vLLMParameter, loaded_weight):
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/linear.py:516]: Let's be more explicit about what this case is. We get `no_shard_id` when the MLP and QKV are fused on disk. Let's also include a link to an example model.

- Example model: https://huggingface.co/microsoft/Phi-3-mini-4k-instruct

File: vllm/model_executor/layers/linear.py [hunk 4]
```diff
@@ -471,6 +493,81 @@ def weight_loader(self,
         assert param_data.shape == loaded_weight.shape
         param_data.copy_(loaded_weight)
 
+    def _default_loading(self, param: vLLMParameter, param_data: torch.Tensor,
+                         loaded_weight: torch.Tensor, loaded_shard_id: int):
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/linear.py:497]: Instead of `_default_loading`, we could call this `load_weight_tensor`, since this function holds the logic for loading the weight (as opposed to the scale or metadata in quantized case)

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_unquantized.py [hunk 5]
```diff
@@ -24,15 +23,14 @@ def create_weights(self, layer: torch.nn.Module,
                        params_dtype: torch.dtype, weight_loader: Callable,
                        **kwargs):
 
-        weight = Parameter(torch.empty(sum(output_partition_sizes),
-                                       input_size_per_partition,
-                                       device="cuda",
-                                       dtype=params_dtype),
-                           requires_grad=False)
+        weight = vLLMParameter(data=torch.empty(sum(output_partition_sizes),
+                                                input_size_per_partition,
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_unquantized.py:27]: I think this needs `use_col_loading=True`

File: vllm/model_executor/parameter.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    def __new__(cls,
+                data: torch.Tensor,
+                requires_grad: Optional[bool] = False,
+                **kwargs):
+
+        return super().__new__(cls, data=data, requires_grad=requires_grad)
+
+    def __init__(self,
+                 data: torch.Tensor,
+                 requires_grad: Optional[bool] = False,
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/parameter.py:26]: `requires_grad` will always be false, so let's just set it to False everywhere + remove the argument

**Review by robertgshaw2-redhat — COMMENTED** [review]
moving comments down

File: vllm/model_executor/parameter.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+    @property
+    def packed_factor(self):
+        return self._packed_factor
+
+    @property
+    def marlin_tile(self):
+        return self._marlin_tile
+
+    def _adjust_marlin_shard(self, shard_size, shard_offset):
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/parameter.py:152]: I think we should rename these:
- `_adjust_shard_indexes_for_marlin`
- `_adjust_shard_indexes_for_packing`

File: vllm/model_executor/layers/linear.py [hunk 8]
```diff
@@ -538,6 +635,109 @@ def __init__(self,
...
+        else:
+            shard_id = tp_rank // self.num_kv_head_replicas
+
+        loaded_weight.narrow(param.output_dim, shard_id * shard_size,
+                             shard_size)
+
+        return param_data, loaded_weight
+
+    def _load_no_shard_id(self, param: vLLMParameter, loaded_weight):
+        shard_offsets = [
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/linear.py:680]: can we re-use `_get_shard_offset_mapping` and `_get_shard_size_mapping` here?

File: vllm/model_executor/layers/linear.py [hunk 9]
```diff
@@ -538,6 +635,109 @@ def __init__(self,
...
+
+    def _get_shard_size_mapping(self, loaded_shard_id: str):
+        shard_size_mapping = {
+            "q": self.num_heads * self.head_size,
+            "k": self.num_kv_heads * self.head_size,
+            "v": self.num_kv_heads * self.head_size,
+        }
+        return shard_size_mapping.get(loaded_shard_id)
+
+    def _default_loading(self, param: vLLMParameter, param_data: torch.Tensor,
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/linear.py:655]: I think we should move `_default_loading` into `vLLMParameter`

Then, the function signature would look like:

```python
class vLLMParameter:
    def load_merged_column_parallel_linear(layer: MergedColumnParallelLinear, loaded_weight: torch.tensors):
          pass
    def load_row_parallel_linear(layer: RowParallelLinear, loaded_weight: torch.tensor):
         pass
    ....
```

Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/linear.py:655]: Then, with this update, in weight_loaderv2 instead of:


```python
if param.use_column_loading:
            param_data, loaded_weight = self._default_loading(
                param=param,
                param_data=param_data,
                loaded_weight=loaded_weight,
                loaded_shard_id=loaded_shard_id)
elif param.use_metadata_loading:  # What case is this?
            shard_size = loaded_weight.shape[0]
            shard_index = ["q", "k", "v"].index(loaded_shard_id)
            param_data = param_data.narrow(0, shard_index * shard_size,
                                           shard_size)
elif param.use_col_shard_split:
      param_data, loaded_weight = param.col_shard_splitter(
            param_data=param_data,
            loaded_weight=loaded_weight,
            shard_id=loaded_shard_id)
```

It would just be:

```python
param_data, loaded_weight = param.loaded_merged_column_parallel_linear(
       layer=self,
       loaded_weight=loaded_weight)
```

This would better encapsulate the logic and simplify the `linear.py` file

**Comment by robertgshaw2-neuralmagic** [comment]
This is much better.

I still think we have too much tied logic between `linear.py` and `vLLMParameter` as we still branching on booleans in `vLLMParameter` in `linear.py` 

I think the following two changes would make a better interface.

### Remove `use_col_loading`, `use_row_loading`, `use_col_shard_indexer`

We currently have a series of booleans in `vLLMParameter` (`use_col_loading`, `use_row_loading`, and `use_col_shard_split`). We then do if/else on these inside `linear.py`. This is still pretty confusing to follow.

I think that instead of using these booleans, we should just create 4 separate `classes`
- `WeightParameter`
- `GroupedScaleParameter`
- `ChannelwiseScaleParameter`
- `PerTensorScaleParameter`

These have the following:
- The weights have use_col_loading=True, use_row_loading=True
- The grouped scales have use_col_loading=True, use_row_loading=True
- The channelwise scales have use_col_loading=True, use_row_loading=False
- The per-tensor scales have use_col_shard_split=True


This will make it more explicit about what case we are in and why we are in each state, rather than implicitly via the booleans we have now.

### Move `_default_loading` into vLLMParameter classes

Function signatures would look like:

```bash
class vLLMParameter:
    def load_merged_column_parallel_linear(layer: MergedColumnParallelLinear, loaded_weight: torch.tensors):
          pass
    def load_row_parallel_linear(layer: RowParallelLinear, loaded_weight: torch.tensor):
         pass
    ....
```

Then, with this update, in weight_loaderv2 instead of:


```python
if param.use_column_loading:
            param_data, loaded_weight = self._default_loading(
                param=param,
                param_data=param_data,
                loaded_weight=loaded_weight,
                loaded_shard_id=loaded_shard_id)
elif param.use_metadata_loading:  # What case is this?
            shard_size = loaded_weight.shape[0]
            shard_index = ["q", "k", "v"].index(loaded_shard_id)
            param_data = param_data.narrow(0, shard_index * shard_size,
                                           shard_size)
elif param.use_col_shard_split:
      param_data, loaded_weight = param.col_shard_splitter(
            param_data=param_data,
            loaded_weight=loaded_weight,
            shard_id=loaded_shard_id)
```

It would just be:

```python
param_data, loaded_weight = param.loaded_merged_column_parallel_linear(
       layer=self,
       loaded_weight=loaded_weight)
```

This would better encapsulate the logic and simplify the `linear.py` file

File: vllm/model_executor/parameter.py [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+
+    def __init__(self,
+                 data: torch.Tensor,
+                 requires_grad: Optional[bool] = False,
+                 use_row_loading: Optional[bool] = False,
+                 use_col_loading: Optional[bool] = False,
+                 input_dim: Optional[int] = None,
+                 output_dim: Optional[int] = None,
+                 weight_loader: Optional[Callable] = None,
+                 packed: Optional[bool] = False,
```
[the remark this answers is not in the document]
Comment (dsikka) [reply vllm/model_executor/parameter.py:32]: added as an easy attribute to check if the parameter is packed (and therefore has the relevant parameter properties). Could also just check if isinstance().

File: vllm/model_executor/layers/linear.py (hunk 8 above)
Comment (dsikka) [reply vllm/model_executor/layers/linear.py:680]: Reuse from where?

File: vllm/model_executor/layers/linear.py (hunk 9 above)
Comment (dsikka) [reply vllm/model_executor/layers/linear.py:655]: I dont agree with moving this logic to the parameter.

The loading done is heavily dependent on the layer itself. At that point, we'll end up overloading the parameter with a lot of layer-specific logic.

Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/linear.py:655]: Yeah, I think the parameters should have:
- `load_merged_column_parallel_linear()`
- `load_qkv_column_parallel_linear()`
- `load_row_parallel_linear()`
- `load_column_parallel_linear()`

File: vllm/model_executor/layers/linear.py [hunk 11]
```diff
@@ -538,6 +613,81 @@ def __init__(self,
...
+                                                       shard_size)
+            self.weight_loader_v2(param, loaded_weight_shard, shard_id)
+
+    def weight_loader_v2(self,
+                         param: vLLMParameter,
+                         loaded_weight: torch.Tensor,
+                         loaded_shard_id: Optional[str] = None):
+
+        param_data = param.data
+        if loaded_shard_id is None:  # special case for certain models
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/linear.py:666]: Let's be more explicit about what this case is. We get no_shard_id when the MLP and QKV are fused on disk. Maybe this function could be called load_fused_module_from_checkpoint

Let's also include a link to an example model.

Example model: https://huggingface.co/microsoft/Phi-3-mini-4k-instruct

File: vllm/model_executor/layers/linear.py [hunk 12]
```diff
@@ -471,6 +494,58 @@ def weight_loader(self,
...
+            self._load_no_shard_id(param, loaded_weight)
+            return
+
+        assert loaded_shard_id < len(self.output_sizes)
+
+        tp_size = get_tensor_model_parallel_world_size()
+        shard_offset = sum(self.output_sizes[:loaded_shard_id]) // tp_size
+        shard_size = self.output_sizes[loaded_shard_id] // tp_size
+
+        if isinstance(param, COLUMN_PARALLEL_LOAD_SUPPORTED):
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/linear.py:538]: Could this just be a no-op for the parameters that do not need reshaping for `COLUMN_PARALLEL_LOAD`

File: vllm/model_executor/layers/linear.py [hunk 13]
```diff
@@ -538,6 +613,81 @@ def __init__(self,
...
+             (self.total_num_heads + self.total_num_kv_heads) * self.head_size,
+             self.total_num_kv_heads * self.head_size),
+        ]
+
+        for shard_id, shard_offset, shard_size in shard_offsets:
+            # Special case for Quantization.
+            # If quantized, we need to adjust the offset and size to account
+            # for the packing.
+
+            if isinstance(param, PackedvLLMParameter
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/linear.py:650]: Can you add a comment about why we need to adjust if `packed_dim==output_dim`?

File: vllm/model_executor/layers/linear.py [hunk 14]
```diff
@@ -538,6 +613,81 @@ def __init__(self,
...
+                return
+            self._load_no_shard_id(param, loaded_weight)
+            return
+
+        assert loaded_shard_id in ["q", "k", "v"]
+
+        shard_offset = self._get_shard_offset_mapping(loaded_shard_id)
+        shard_size = self._get_shard_size_mapping(loaded_shard_id)
+
+        if isinstance(param, COLUMN_PARALLEL_LOAD_SUPPORTED):
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/linear.py:679]: Could `load_qkv_weight` just be a no-op for the parameters that do not need reshaping for COLUMN_PARALLEL_LOAD rather than checking this condition here?

File: vllm/model_executor/layers/linear.py [hunk 15]
```diff
@@ -776,6 +931,19 @@ def weight_loader(self, param: Parameter, loaded_weight: torch.Tensor):
         assert param_data.shape == loaded_weight.shape
         param_data.copy_(loaded_weight)
 
+    def weight_loader_v2(self, param: vLLMParameter,
+                         loaded_weight: torch.Tensor):
+
+        if isinstance(param, ROW_PARALLEL_LOAD_SUPPORTED):
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/linear.py:937]: Could `load_row_parallel_weight` just be a no-op instead of checking `ROW_PARALLEL_LOAD_SUPPORTED`

File: vllm/model_executor/parameter.py [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+
+    def __init__(self, output_dim: int, **kwargs):
+        self._output_dim = output_dim
+        super().__init__(**kwargs)
+
+    @property
+    def output_dim(self):
+        return self._output_dim
+
+    def load_column_parallel_weight(self, loaded_weight: torch.Tensor,
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/parameter.py:47]: These functions do not `load` the weights but rather `reshape`

We should move the 

```python
assert param_data.shape == loaded_weight.shape
param_data.copy_(loaded_weight)
```

Into this

Comment (dsikka) [reply vllm/model_executor/parameter.py:47]: No because we use this assertion and copy for everything else that doesnt require sharding or param/loaded weight manipulation 

Such as our input scales for activation quantization

Comment (dsikka) [reply vllm/model_executor/parameter.py:47]: We can update the names of the parameter functions to use a better verb than `load`

File: vllm/model_executor/layers/linear.py (hunk 15 above)
Comment (dsikka) [reply vllm/model_executor/layers/linear.py:937]: yes - tradeoff would be more code to the parameter classes

File: vllm/model_executor/layers/linear.py (hunk 14 above)
Comment (dsikka) [reply vllm/model_executor/layers/linear.py:679]: yes - tradeoff would be more code to the parameter classes

File: vllm/model_executor/parameter.py [hunk 17]
```diff
@@ -0,0 +1,191 @@
+from typing import Callable, List, Optional, Tuple, Union
+
+import torch
+from torch.nn import Parameter
+
+from vllm.distributed import get_tensor_model_parallel_rank
+from vllm.logger import init_logger
+
+__all__ = [
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/parameter.py:9]: This is a complete nit, but I don't think its best practice to have a lowercase at the start of Python classes.

Potential naming scheme:

```bash
- `ModelParameter`
- `WeightParameter`
- `PackedWeigthParameter`
- `PerGroupScaleParameter`
- `PerChannelScaleParameter`
- `PerTensorScaleParameter`
```

Comment (dsikka) [reply vllm/model_executor/parameter.py:9]: I like the vLLM in the name - we can add something to the from
i.e `ModelvLLMParameter`

File: vllm/model_executor/parameter.py [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+
+        tp_rank = get_tensor_model_parallel_rank()
+        param_data = param_data.narrow(self.output_dim, shard_offset,
+                                       shard_size)
+        loaded_weight = loaded_weight.narrow(self.output_dim,
+                                             tp_rank * shard_size, shard_size)
+        return param_data, loaded_weight
+
+    def load_qkv_weight(self, param_data: torch.Tensor,
+                        loaded_weight: torch.Tensor, num_heads: int,
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/parameter.py:73]: We should not have to pass `param_data` to this function since `param_data=self.data`

Comment (dsikka) [reply vllm/model_executor/parameter.py:73]: yup, typo

File: vllm/model_executor/parameter.py [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+__all__ = [
+    "vLLMParameter", "PackedvLLMParameter", "PerTensorScaleParameter",
+    "ModelWeightParameter", "ChannelQuantScaleParameter",
+    "GroupQuantScaleParameter"
+]
+
+logger = init_logger(__name__)
+
+
+class vLLMParameter(Parameter):
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/parameter.py:18]: Can this be an abstract class that exposes:
- `load_column_parallel_weight`
- `load_qkv_weight`
- `load_merged_column_parallel_weight`
- `load_row_column_parallel_weight`

File: vllm/model_executor/parameter.py (hunk 16 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/parameter.py:47]: But input scales will be `PerTensorScale` no?

Isn't the idea that all parameters will call `load_xxx_`?

File: vllm/model_executor/parameter.py (hunk 19 above)
Comment (dsikka) [reply vllm/model_executor/parameter.py:18]: I think no-ops over abstract. We want to return the loaded weight and parameter data unchanged in the base class

File: vllm/model_executor/parameter.py (hunk 16 above)
Comment (dsikka) [reply vllm/model_executor/parameter.py:47]: No it won't be
We also have other parameters that dont require the manipulation - such as the weight_shape for packed cases

File: vllm/model_executor/parameter.py (hunk 17 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/parameter.py:9]: `TensorParallelParam`?

File: vllm/model_executor/parameter.py (hunk 19 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/parameter.py:18]: See my updated note^

Perhaps we could just put the copying there?

File: vllm/model_executor/parameter.py [hunk 20]
```diff
@@ -0,0 +1,191 @@
...
+
+        self._ignore_warnings = True
+        self._weight_loader = weight_loader
+
+    @property
+    def weight_loader(self):
+        return self._weight_loader
+
+
+class _ColumnvLLMParameter(vLLMParameter):
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/parameter.py:37]: I find multiple levels of inheritance to be hard to follow.

There is definitely some re-used code here though. 

Perhaps we could have some floating functions that are shared across the subclasses?

File: vllm/model_executor/parameter.py (hunk 17 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/parameter.py:9]: `BasevLLMParameter` would work well

File: vllm/model_executor/parameter.py (hunk 16 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/parameter.py:47]: Sure, and `BasevLLMParameter` will have the passthrough right?

Also - why doesn't InputScale have the same manipulation? We still have 3 scales but the cutlass kernels only accept 1 scale no?

File: vllm/model_executor/layers/linear.py (hunk 8 above)
Comment (dsikka) [reply vllm/model_executor/layers/linear.py:632]: slightly different mappings

File: vllm/model_executor/parameter.py [hunk 21]
```diff
@@ -0,0 +1,277 @@
...
+
+    def load_qkv_weight(self, *args, **kwargs):
+        self._col_shard_splitter(*args, **kwargs)
+
+    def load_column_parallel_weight(self, *args, **kwargs):
+        self._col_shard_splitter(*args, **kwargs)
+
+    def _col_shard_splitter(self, loaded_weight: torch.Tensor,
+                            shard_id: Union[str, int], **kwargs):
+        """
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/parameter.py:216]: Can we call this `load_into_shard_idx`?

File: vllm/model_executor/parameter.py [hunk 22]
```diff
@@ -0,0 +1,277 @@
...
+    def load_merged_column_weight(self, loaded_weight: torch.Tensor, **kwargs):
+        self._assert_and_load(loaded_weight)
+
+    def load_qkv_weight(self, loaded_weight: torch.Tensor, **kwargs):
+        self._assert_and_load(loaded_weight)
+
+
+class _ColumnvLLMParameter(BasevLLMParameter):
+    """
+    Private class defining weight loading functionality
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/parameter.py:64]: This comment is a bit confusing

Its more --> "needs to handle sharding in QKV or MLP case"

**Comment by robertgshaw2-neuralmagic** [comment]
Couple nits but LGTM

**Review by robertgshaw2-redhat — APPROVED** [review]

File: vllm/model_executor/parameter.py (hunk 22 above)
Comment (dsikka) [reply vllm/model_executor/parameter.py:64]: There's also the generic case for column parallelism, not specific to QKV or MLP

**Comment by comaniac** [comment]
At the first glance I didn't see `linear.py` or other linear method begin simplified except for compressed tensors? `linear.py` becomes even more complicate actually...is that because you want v1 and v2 at the same time?

**Comment by robertgshaw2-neuralmagic** [comment]
@comaniac weight_loaderv1 will be removed once we finish converting the rest of the integrations over. We thought it made sense to do it one integration at a time to limit the scope of each PR

**Comment by comaniac** [comment]
Per offline discussion we will maintain a list for unsupported quant methods, and we could remove v2 once the list becomes empty. All new quant methods should just go with v2.

**Comment by dsikka** [comment]
FYI: 

The functionality covered by `create_per_channel_scale_param` and `create_per_tensor_scale_param` is done by the newly added parameters so these functions are not being used in the compressed-tensors schemes.

Will be updated to use the functions once we adapt the refactor to other quant configs.

File: vllm/model_executor/layers/linear.py [hunk 23]
```diff
@@ -284,14 +286,20 @@ def __init__(self,
...
             layer=self,
             input_size_per_partition=self.input_size,
             output_partition_sizes=self.output_partition_sizes,
             input_size=self.input_size,
             output_size=self.output_size,
             params_dtype=self.params_dtype,
-            weight_loader=self.weight_loader,
+            weight_loader=(self.weight_loader_v2 if isinstance(
+                self.quant_method, CompressedTensorsLinearMethod) else
+                           self.weight_loader),
```
Comment (comaniac) [inline vllm/model_executor/layers/linear.py:302]: Per offline discussion with @robertgshaw2-neuralmagic, we may want to maintain a list of unsupported methods so that we could easily track when could we deprecate weight loader v1.

Comment (dsikka) [reply vllm/model_executor/layers/linear.py]: @comaniac Updated.

**Review by comaniac — APPROVED** [review]
Sorry for the late review. Overall LGTM so approve to unblock this PR and follow-up tasks.

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_unquantized.py [hunk 24]
```diff
@@ -24,22 +23,24 @@ def get_min_capability(cls) -> int:
         return 70
 
     def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
-        pass
+        layer.weight = torch.nn.Parameter(layer.weight.data,
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_unquantized.py:26]: Can you add a comment about why this needs to happen (e.g. for torch.compile)?

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w4a16_24.py [hunk 25]
```diff
@@ -35,7 +38,11 @@ def get_min_capability(cls) -> int:
         return 80
 
     def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
-        pass
+        layer.weight_packed = Parameter(layer.weight_packed.data,
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w4a16_24.py:41]: add comment for why this is needed

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a16_fp8.py [hunk 26]
```diff
@@ -40,11 +41,17 @@ def process_weights_after_loading(self, layer) -> None:
                                                     layer.logical_widths)
             layer.weight_scale = torch.nn.Parameter(ws_channelwise,
                                                     requires_grad=False)
+        else:
+            layer.weight_scale = torch.nn.Parameter(layer.weight_scale.data,
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a16_fp8.py:45]: add comment for why this is needed

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a16_fp8.py [hunk 27]
```diff
@@ -60,35 +67,39 @@ def create_weights(self, layer: torch.nn.Module, input_size: int,
...
-                output_partition_sizes, **layer_kwargs)
+            weight_scale = PerTensorScaleParameter(data=torch.empty(
+                len(output_partition_sizes), dtype=torch.float32),
+                                                   weight_loader=weight_loader)
         else:
             raise ValueError(
                 f"Unsupported weight strategy={self.strategy}, "
                 f"supported strategies are {SUPPORTED_STRATEGIES}")
+
+        weight_scale[:] = torch.finfo(torch.float32).min
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a16_fp8.py:95]: Do you think this should be performed in the init function for `PerTensorScaleParam`?

Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a16_fp8.py:95]: Perhaps it is better outside

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py [hunk 28]
```diff
@@ -46,6 +46,8 @@ def process_weights_after_loading(self, layer) -> None:
         elif self.strategy == QuantizationStrategy.CHANNEL:
             weight = layer.weight
             layer.weight = Parameter(weight.t(), requires_grad=False)
+            layer.weight_scale = Parameter(layer.weight_scale.data,
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py:49]: add comment for why this is needed

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py [hunk 29]
```diff
@@ -66,32 +68,39 @@ def create_weights(self, layer: torch.nn.Module,
...
         layer.register_parameter("weight", weight)
-        set_weight_attrs(weight, {
-            "input_dim": 1,
-            "output_dim": 0,
-            "weight_loader": weight_loader,
-        })
 
         # WEIGHT SCALE
-        layer_kwargs = {"weight_loader": weight_loader}
+        # TODO: update create_xxx_parameter functions to return
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py:81]: Is this still a TODO?

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a16_fp8.py (hunk 27 above)
Comment (dsikka) [reply vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a16_fp8.py:95]: @robertgshaw2-neuralmagic from what I understand, this is only needed for fp8, not all PerTensorScaleParams?

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py [hunk 30]
```diff
@@ -154,10 +144,15 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
 
         # No zero-point
         layer.weight_zp = marlin_make_empty_g_idx(device)
+        # Update for kernel
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py:147]: Add a comment for why this is needed

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py (hunk 29 above)
Comment (dsikka) [reply vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py:81]: Yes. We're not using the create_xxx_parameter methods here as they are used in places outside of compressed_tensors (e.g fp8). As a follow-up, once we've updated other quantization methods to use these new parameters, we can update the create_xx_parameter functions to return the vLLMParameters

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py [hunk 31]
```diff
@@ -63,11 +61,12 @@ def create_weights(self, layer: torch.nn.Module, input_size: int,
...
                        params_dtype: torch.dtype, weight_loader: Callable,
                        **kwargs):
+
         output_size_per_partition = sum(output_partition_sizes)
 
         # If group_size is -1, we are in channelwise case.
-        channelwise = (self.group_size == -1)
-        group_size = input_size if channelwise else self.group_size
+        channelwise = self.group_size == -1
+        group_size = self.group_size if self.group_size != -1 else input_size
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py:69]: why change this?

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py [hunk 32]
```diff
@@ -230,14 +225,16 @@ def _get_scheme_from_parts(
                     group_size=weight_quant.group_size)
 
         # Detect If Activation Quantization.
+        # TODO @dsikka: clean-up conditions
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:228]: Is this still a TODO?

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py [hunk 33]
```diff
@@ -230,14 +225,16 @@ def _get_scheme_from_parts(
...
         if is_activation_quantization_format(self.quant_format):
             if self._is_fp8_w8a8(weight_quant, input_quant):
                 is_fp8_w8a8_supported = self._check_scheme_supported(
                     CompressedTensorsW8A8Fp8.get_min_capability(), error=False)
                 if is_fp8_w8a8_supported:
                     return CompressedTensorsW8A8Fp8(
                         strategy=weight_quant.strategy,
-                        is_static_input_scheme=(not input_quant.dynamic))
+                        is_static_input_scheme=(input_quant
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:236]: Why is this changing? Won't be always have `input_quant` if `is_activation_quantization_format`?

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py (hunk 32 above)
Comment (dsikka) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:228]: Yes. General follow-up on the state of these conditions

**Comment by robertgshaw2-neuralmagic** [comment]
Make sure to unblock the multi-gpu A100 model correctness tests. Nice job!

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py (hunk 31 above)
Comment (dsikka) [reply vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_wNa16.py:69]: the second condition is just clearer as what the group_size is and why

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py (hunk 33 above)
Comment (dsikka) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors.py:236]: just an extra check the activation config details aren't None/parsed correctly.
